scripts/train_price_prediction_model.py

```diff
@@ -0,0 +1,115 @@
+            rmse_score = np.sqrt(mse_score)
+            metrics_for_all_folds = np.append(
+                metrics_for_all_folds,
+                np.array([(mae_score, mse_score, rmse_score)], dtype=[("mae", "float32"), ("mse", "float32"), ("rmse", "float32")])
+                )
+
+        loss = metric_result('rmse', metrics_for_all_folds)['mean_val']
+        results = { 'loss': loss, 'status': STATUS_OK }
+        y_interquartile_range = y.describe()['75%'] - y.describe()['25%']
+        for metric in metrics_for_all_folds.dtype.names:
+            results[metric] = metric_result(metric, metrics_for_all_folds)
+            results[metric+'_percentage_of_iqr'] = 100*results[metric]['mean_val']/y_interquartile_range
+        return results
+    return objective
+
+def optimize_hyperparams(X: pd.DataFrame, y: pd.DataFrame) -> Tuple[dict, Trials]:
+    space = {
+        'verbosity': 0,
+        'learning_rate': hp.choice('learning_rate', np.arange(0.001, 0.99, 0.001)),
+        'max_depth': hp.choice('max_depth', np.arange(2, 10, 1, dtype=int)),
+        'min_child_weight': hp.choice('min_child_weight', np.arange(1, 10, 1, dtype=int)),
+        'colsample_bytree': hp.choice('colsample_bytree', np.arange(0.1, 1.0, 0.1)),
+        'subsample': hp.uniform('subsample', 0.5, 1),
+        'n_estimators': hp.choice('n_estimators', np.arange(1, 200, 1))
+    }
+    trials = Trials()
+    best = fmin(
+        fn=kfold_scorer(X, y),
+        space=space,
+        algo=tpe.suggest,
+        max_evals=100,
+        trials=trials
+    )
+    best_params = space_eval(space, best)
+    return best_params, trials
+    
+def main(X: pd.DataFrame, y: pd.DataFrame) -> Tuple[XGBRegressor, dict]:
+    best_params, trials = optimize_hyperparams(X, y)
+    model = XGBRegressor(**best_params)
+    model.fit(X.values, y.values)
+
+    int64_to_int = lambda x: int(x) if type(x) == np.int64 else x
+    best_params = {k:int64_to_int(v) for k, v in best_params.items()} # np.int64 objects are not JSON serializable
+    feature_importances = sort_dict_by_value(
+        {feature: importance for feature, importance in zip(X.columns.to_list(), model.feature_importances_.tolist())}
+    )
+    model_details = {
+        'best_params': best_params,
+        'best_trial': trials.best_trial['result'],
+        'feature_importances': feature_importances
+    }
+    return model, model_details
+
+
+
+if __name__ == '__main__':
+    training_data = pd.read_csv(PROJ_DIR+'/data/training_data.csv')
+    X = training_data[features]
+    y = training_data['price'].apply(lambda x: np.log(1+x)) # Changing target variable to log of price cause price is massively right skewed
+    model, model_details = main(X, y)
+    pprint(model_details, sort_dicts=False)
+    save_as_pickle(obj=model, filepath=PROJ_DIR+'/models/price_prediction_model.pkl')
+    save_as_json(obj=model_details, filepath=PROJ_DIR+'/models/price_prediction_metrics.json')
```

Comment:
ayy big man teach meee how to build models like this

Reply:
It's a light oneee 👀

Reply:
Nah bmt though